File: compression_pipeline.py
import zlib
import pickle
import sqlite3
import time
import json
from pathlib import Path
from typing import List, Tuple, Dict, Any, Optional
from dataclasses import dataclass, asdict
from llm_ranker import LLMRanker
from ranker_cache import RankerCache


@dataclass
class CompressionResult:
    """Results from a compression experiment."""
    method: str
    original_size: int
    compressed_size: int
    compression_ratio: float
    compression_time: float
    decompression_time: float
    model_name: Optional[str] = None
    context_length: Optional[int] = None
    text_sample: str = ""  # First 100 chars for reference
    
    @property
    def compression_percentage(self) -> float:
        """Compression as percentage reduction."""
        return (1 - self.compression_ratio) * 100


class CompressionPipeline:
    """Main compression pipeline with database storage for results."""
    
    def __init__(self, db_path: str = "compression_results.db"):
        """Initialize the compression pipeline.
        
        Args:
            db_path: Path to SQLite database for storing results
        """
        self.db_path = db_path
        self.ranker: Optional[LLMRanker] = None
        self.cache = RankerCache(db_path)
        self._setup_database()
    
    def _setup_database(self):
        """Initialize the SQLite database for storing results."""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Create compression results table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS compression_results (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                timestamp TEXT NOT NULL,
                method TEXT NOT NULL,
                original_size INTEGER NOT NULL,
                compressed_size INTEGER NOT NULL,
                compression_ratio REAL NOT NULL,
                compression_time REAL NOT NULL,
                decompression_time REAL NOT NULL,
                model_name TEXT,
                context_length INTEGER,
                text_sample TEXT,
                metadata TEXT
            )
        """)
        
        conn.commit()
        conn.close()
    
    def _get_ranker(self, model_name: str = "gpt2", max_context_length: Optional[int] = None) -> LLMRanker:
        """Get or create LLM ranker instance."""
        if (self.ranker is None or 
            self.ranker.model_name != model_name or 
            self.ranker.max_context_length != max_context_length):
            self.ranker = LLMRanker(
                model_name=model_name,
                max_context_length=max_context_length
            )
        return self.ranker
    
    def get_token_ranks_cached(self, text: str, ranker: LLMRanker) -> List[int]:
        """Get token ranks with caching."""
        return self.cache.get_token_ranks_cached(text, ranker)
    
    def get_string_from_token_ranks_cached(self, ranks: List[int], ranker: LLMRanker) -> str:
        """Get string from token ranks with caching."""
        return self.cache.get_string_from_token_ranks_cached(ranks, ranker)
    
    def compress_with_llm_ranks(
        self, 
        text: str, 
        model_name: str = "gpt2",
        max_context_length: Optional[int] = None
    ) -> Tuple[bytes, CompressionResult]:
        """Compress text using LLM rank encoding + zlib.
        
        Args:
            text: Input text to compress
            model_name: LLM model to use for ranking
            max_context_length: Maximum context length for LLM
            
        Returns:
            Tuple of (compressed_data, compression_result)
        """
        ranker = self._get_ranker(model_name, max_context_length)
        
        # Phase 1: Get token ranks (with caching)
        start_time = time.time()
        ranks = self.get_token_ranks_cached(text, ranker)
        
        # Phase 2: Compress ranks with zlib
        ranks_bytes = pickle.dumps(ranks)
        compressed_data = zlib.compress(ranks_bytes)
        compression_time = time.time() - start_time
        
        # Test decompression
        start_time = time.time()
        decompressed_ranks = pickle.loads(zlib.decompress(compressed_data))
        reconstructed_text = self.get_string_from_token_ranks_cached(decompressed_ranks, ranker)
        decompression_time = time.time() - start_time
        
        # Verify round-trip accuracy
        if reconstructed_text != text:
            raise ValueError("Round-trip compression failed - reconstructed text doesn't match original")
        
        original_size = len(text.encode('utf-8'))
        compressed_size = len(compressed_data)
        
        result = CompressionResult(
            method="llm_ranks_zlib",
            original_size=original_size,
            compressed_size=compressed_size,
            compression_ratio=compressed_size / original_size,
            compression_time=compression_time,
            decompression_time=decompression_time,
            model_name=model_name,
            context_length=max_context_length,
            text_sample=text[:100]
        )
        
        return compressed_data, result
    
    def compress_with_raw_zlib(self, text: str) -> Tuple[bytes, CompressionResult]:
        """Baseline: compress raw text with zlib.
        
        Args:
            text: Input text to compress
            
        Returns:
            Tuple of (compressed_data, compression_result)
        """
        text_bytes = text.encode('utf-8')
        
        # Compression
        start_time = time.time()
        compressed_data = zlib.compress(text_bytes)
        compression_time = time.time() - start_time
        
        # Test decompression
        start_time = time.time()
        decompressed_data = zlib.decompress(compressed_data)
        decompression_time = time.time() - start_time
        
        # Verify round-trip
        if decompressed_data.decode('utf-8') != text:
            raise ValueError("Round-trip compression failed")
        
        original_size = len(text_bytes)
        compressed_size = len(compressed_data)
        
        result = CompressionResult(
            method="raw_zlib",
            original_size=original_size,
            compressed_size=compressed_size,
            compression_ratio=compressed_size / original_size,
            compression_time=compression_time,
            decompression_time=decompression_time,
            text_sample=text[:100]
        )
        
        return compressed_data, result
    
    def compress_with_tokenizer_zlib(
        self, 
        text: str, 
        model_name: str = "gpt2"
    ) -> Tuple[bytes, CompressionResult]:
        """Baseline: tokenize then compress token IDs with zlib.
        
        Args:
            text: Input text to compress
            model_name: Tokenizer model to use
            
        Returns:
            Tuple of (compressed_data, compression_result)
        """
        ranker = self._get_ranker(model_name)
        
        # Tokenize
        start_time = time.time()
        token_ids = ranker.tokenizer.encode(text)
        token_bytes = pickle.dumps(token_ids)
        compressed_data = zlib.compress(token_bytes)
        compression_time = time.time() - start_time
        
        # Test decompression
        start_time = time.time()
        decompressed_tokens = pickle.loads(zlib.decompress(compressed_data))
        reconstructed_text = ranker.tokenizer.decode(decompressed_tokens, skip_special_tokens=True)
        decompression_time = time.time() - start_time
        
        original_size = len(text.encode('utf-8'))
        compressed_size = len(compressed_data)
        
        result = CompressionResult(
            method="tokenizer_zlib",
            original_size=original_size,
            compressed_size=compressed_size,
            compression_ratio=compressed_size / original_size,
            compression_time=compression_time,
            decompression_time=decompression_time,
            model_name=model_name,
            text_sample=text[:100]
        )
        
        return compressed_data, result
    
    def run_compression_benchmark(
        self, 
        text: str, 
        model_name: str = "gpt2",
        max_context_length: Optional[int] = None,
        save_to_db: bool = True
    ) -> Dict[str, CompressionResult]:
        """Run all compression methods on text and return results.
        
        Args:
            text: Input text to compress
            model_name: LLM model to use
            max_context_length: Maximum context length for LLM
            save_to_db: Whether to save results to database
            
        Returns:
            Dictionary mapping method names to results
        """
        results = {}
        
        # Run all compression methods
        try:
            _, results["llm_ranks_zlib"] = self.compress_with_llm_ranks(
                text, model_name, max_context_length
            )
        except Exception as e:
            print(f"LLM ranks compression failed: {e}")
            results["llm_ranks_zlib"] = None
        
        try:
            _, results["raw_zlib"] = self.compress_with_raw_zlib(text)
        except Exception as e:
            print(f"Raw zlib compression failed: {e}")
            results["raw_zlib"] = None
        
        try:
            _, results["tokenizer_zlib"] = self.compress_with_tokenizer_zlib(text, model_name)
        except Exception as e:
            print(f"Tokenizer zlib compression failed: {e}")
            results["tokenizer_zlib"] = None
        
        # Save to database if requested
        if save_to_db:
            for method, result in results.items():
                if result is not None:
                    self._save_result_to_db(result)
        
        return results
    
    def _save_result_to_db(self, result: CompressionResult):
        """Save a compression result to the database."""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            INSERT INTO compression_results (
                timestamp, method, original_size, compressed_size, compression_ratio,
                compression_time, decompression_time, model_name, context_length, 
                text_sample, metadata
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            time.strftime('%Y-%m-%d %H:%M:%S'),
            result.method,
            result.original_size,
            result.compressed_size,
            result.compression_ratio,
            result.compression_time,
            result.decompression_time,
            result.model_name,
            result.context_length,
            result.text_sample,
            json.dumps(asdict(result))
        ))
        
        conn.commit()
        conn.close()
    
    def get_results_from_db(self, method: Optional[str] = None) -> List[Dict]:
        """Retrieve compression results from database.
        
        Args:
            method: Filter by compression method (optional)
            
        Returns:
            List of result dictionaries
        """
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        if method:
            cursor.execute("""
                SELECT * FROM compression_results WHERE method = ?
                ORDER BY timestamp DESC
            """, (method,))
        else:
            cursor.execute("""
                SELECT * FROM compression_results ORDER BY timestamp DESC
            """)
        
        columns = [desc[0] for desc in cursor.description]
        results = [dict(zip(columns, row)) for row in cursor.fetchall()]
        
        conn.close()
        return results
    
    def print_comparison_table(self, results: Dict[str, CompressionResult]):
        """Print a formatted comparison table of compression results."""
        if not any(results.values()):
            print("No successful compression results to display.")
            return
        
        print("\n" + "="*80)
        print("COMPRESSION COMPARISON RESULTS")
        print("="*80)
        
        # Table header
        print(f"{'Method':<20} {'Original':<10} {'Compressed':<10} {'Ratio':<8} {'Reduction':<10} {'Time (s)':<10}")
        print("-" * 80)
        
        # Sort by compression ratio (best first)
        sorted_results = sorted(
            [(k, v) for k, v in results.items() if v is not None],
            key=lambda x: x[1].compression_ratio
        )
        
        for method, result in sorted_results:
            print(f"{method:<20} {result.original_size:<10} {result.compressed_size:<10} "
                  f"{result.compression_ratio:<8.3f} {result.compression_percentage:<9.1f}% "
                  f"{result.compression_time:<10.3f}")
        
        print("-" * 80)
        
        # Show best performing method
        if sorted_results:
            best_method, best_result = sorted_results[0]
            print(f"Best compression: {best_method} ({best_result.compression_percentage:.1f}% reduction)")
        
        print("="*80)


def load_dataset_file(filename: str) -> str:
    """Load a text file from the datasets directory."""
    file_path = Path("datasets") / filename
    if not file_path.exists():
        raise FileNotFoundError(f"Dataset file not found: {file_path}")
    
    with open(file_path, 'r', encoding='utf-8') as f:
        return f.read()


def main():
    """Example usage of the compression pipeline."""
    pipeline = CompressionPipeline()
    
    # Test with a simple example
    test_text = "Hello world! This is a test of the compression pipeline. " * 10
    print(f"Testing with sample text ({len(test_text)} characters)")
    
    results = pipeline.run_compression_benchmark(test_text)
    pipeline.print_comparison_table(results)


if __name__ == "__main__":
    main()

File: src/compression-pipeline.js
import { LLMRanker } from './llm-ranker.js';
import { HuffmanRankCompressor } from './huffman-compression.js';
import { RankerCache } from './ranker-cache.js';
import { promisify } from 'util';
import zlib from 'zlib';

// Try to import sqlite3, fall back to mock if not available
let Database;
try {
    const sqlite3 = await import('sqlite3');
    Database = sqlite3.Database;
} catch (error) {
    // Mock Database class for testing without sqlite3
    Database = class MockDatabase {
        constructor() {}
        run() {}
        get(query, params, callback) { callback(null, null); }
        all(query, params, callback) { callback(null, []); }
        close() {}
    };
}

const gzip = promisify(zlib.gzip);
const gunzip = promisify(zlib.gunzip);

/**
 * Results from a compression experiment
 */
export class CompressionResult {
    constructor(options = {}) {
        this.method = options.method || '';
        this.originalSize = options.originalSize || 0;
        this.compressedSize = options.compressedSize || 0;
        this.compressionRatio = options.compressionRatio || 0;
        this.compressionTime = options.compressionTime || 0;
        this.decompressionTime = options.decompressionTime || 0;
        this.modelName = options.modelName || null;
        this.contextLength = options.contextLength || null;
        this.textSample = options.textSample || '';
    }

    /**
     * Get compression as percentage reduction
     * @returns {number} Compression percentage
     */
    get compressionPercentage() {
        return (1 - this.compressionRatio) * 100;
    }
}

/**
 * Main compression pipeline with database storage for results
 */
export class CompressionPipeline {
    /**
     * Initialize the compression pipeline
     * @param {Object} options - Configuration options
     * @param {string} options.dbPath - Path to SQLite database
     * @param {boolean} options.enableCache - Whether to enable caching
     */
    constructor(options = {}) {
        this.dbPath = options.dbPath || 'compression_results.db';
        this.enableCache = options.enableCache !== false;
        
        this.ranker = null;
        this.cache = this.enableCache ? new RankerCache(this.dbPath) : null;
        this.huffmanCompressor = new HuffmanRankCompressor();
        
        this._setupDatabase();
    }

    /**
     * Initialize the SQLite database
     * @private
     */
    _setupDatabase() {
        const db = new Database(this.dbPath);
        
        // Create compression results table
        db.run(`
            CREATE TABLE IF NOT EXISTS compression_results (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                timestamp TEXT NOT NULL,
                method TEXT NOT NULL,
                original_size INTEGER NOT NULL,
                compressed_size INTEGER NOT NULL,
                compression_ratio REAL NOT NULL,
                compression_time REAL NOT NULL,
                decompression_time REAL NOT NULL,
                model_name TEXT,
                context_length INTEGER,
                text_sample TEXT,
                metadata TEXT
            )
        `);
        
        db.close();
    }

    /**
     * Get or create LLM ranker instance
     * @param {string} modelName - Model name
     * @param {number} maxContextLength - Maximum context length
     * @param {number} batchSize - Batch size
     * @returns {LLMRanker} Ranker instance
     * @private
     */
    _getRanker(modelName = 'gpt2', maxContextLength = null, batchSize = 1) {
        if (!this.ranker || 
            this.ranker.modelName !== modelName ||
            this.ranker.maxContextLength !== maxContextLength ||
            this.ranker.batchSize !== batchSize) {
            
            this.ranker = new LLMRanker({
                modelName,
                maxContextLength,
                batchSize
            });
        }
        
        return this.ranker;
    }

    /**
     * Get token ranks with optional caching
     * @param {string} text - Input text
     * @param {LLMRanker} ranker - Ranker instance
     * @returns {Array<number>} Token ranks
     */
    async getTokenRanksCached(text, ranker) {
        if (this.enableCache && this.cache) {
            return await this.cache.getTokenRanksCached(text, ranker);
        } else {
            return await ranker.getTokenRanks(text);
        }
    }

    /**
     * Get string from token ranks with optional caching
     * @param {Array<number>} ranks - Token ranks
     * @param {LLMRanker} ranker - Ranker instance
     * @returns {string} Generated text
     */
    async getStringFromTokenRanksCached(ranks, ranker) {
        if (this.enableCache && this.cache) {
            return await this.cache.getStringFromTokenRanksCached(ranks, ranker);
        } else {
            return await ranker.getStringFromTokenRanks(ranks);
        }
    }

    /**
     * Create a compression result object
     * @param {string} originalText - Original text
     * @param {Buffer} compressedData - Compressed data
     * @param {number} compressionTime - Compression time
     * @param {number} decompressionTime - Decompression time
     * @param {string} methodName - Method name
     * @param {string} modelName - Model name
     * @param {number} contextLength - Context length
     * @returns {CompressionResult} Result object
     * @private
     */
    _createCompressionResult(originalText, compressedData, compressionTime, 
                           decompressionTime, methodName, modelName = null, 
                           contextLength = null) {
        const originalSize = Buffer.byteLength(originalText, 'utf8');
        const compressedSize = compressedData.length;
        
        return new CompressionResult({
            method: methodName,
            originalSize,
            compressedSize,
            compressionRatio: compressedSize / originalSize,
            compressionTime,
            decompressionTime,
            modelName,
            contextLength,
            textSample: originalText.substring(0, 100)
        });
    }

    /**
     * Compress text using LLM rank encoding + gzip
     * @param {string} text - Input text
     * @param {string} modelName - Model name
     * @param {number} maxContextLength - Maximum context length
     * @returns {Object} {compressedData: Buffer, result: CompressionResult}
     */
    async compressWithLLMRanks(text, modelName = 'gpt2', maxContextLength = null) {
        const ranker = this._getRanker(modelName, maxContextLength);
        
        // Phase 1: Get token ranks
        const startTime = Date.now();
        const ranks = await this.getTokenRanksCached(text, ranker);
        
        // Phase 2: Compress ranks with gzip
        const ranksBuffer = Buffer.from(JSON.stringify(ranks));
        const compressedData = await gzip(ranksBuffer);
        const compressionTime = (Date.now() - startTime) / 1000;
        
        // Test decompression
        const decompStartTime = Date.now();
        const decompressedRanks = JSON.parse((await gunzip(compressedData)).toString());
        const reconstructedText = await this.getStringFromTokenRanksCached(decompressedRanks, ranker);
        const decompressionTime = (Date.now() - decompStartTime) / 1000;
        
        // Verify round-trip accuracy
        if (reconstructedText !== text) {
            throw new Error('Round-trip compression failed - reconstructed text doesn\'t match original');
        }
        
        const result = this._createCompressionResult(
            text, compressedData, compressionTime, decompressionTime,
            'llm_ranks_gzip', modelName, maxContextLength
        );
        
        return { compressedData, result };
    }

    /**
     * Compress text using LLM rank encoding + Huffman coding
     * @param {string} text - Input text
     * @param {string} modelName - Model name
     * @param {number} maxContextLength - Maximum context length
     * @returns {Object} {compressedData: Buffer, result: CompressionResult}
     */
    async compressWithLLMRanksHuffman(text, modelName = 'gpt2', maxContextLength = null) {
        const ranker = this._getRanker(modelName, maxContextLength);
        
        // Phase 1: Get token ranks
        const ranks = await this.getTokenRanksCached(text, ranker);
        
        // Phase 2: Compress ranks with Huffman coding
        const { compressedData, compressionTime } = 
            this.huffmanCompressor.compressRanksWithTiming(ranks, 'basic');
        
        // Phase 3: Test decompression
        const { decompressedRanks, decompressionTime } = 
            this.huffmanCompressor.decompressRanksWithTiming(compressedData, 'basic');
        const reconstructedText = await this.getStringFromTokenRanksCached(decompressedRanks, ranker);
        
        // Verify round-trip accuracy
        if (reconstructedText !== text) {
            throw new Error('Round-trip compression failed - reconstructed text doesn\'t match original');
        }
        
        const result = this._createCompressionResult(
            text, compressedData, compressionTime, decompressionTime,
            'llm_ranks_huffman', modelName, maxContextLength
        );
        
        return { compressedData, result };
    }

    /**
     * Compress text using LLM rank encoding + Huffman coding with Zipf distribution
     * @param {string} text - Input text
     * @param {string} modelName - Model name
     * @param {number} maxContextLength - Maximum context length
     * @returns {Object} {compressedData: Buffer, result: CompressionResult}
     */
    async compressWithLLMRanksHuffmanZipf(text, modelName = 'gpt2', maxContextLength = null) {
        const ranker = this._getRanker(modelName, maxContextLength);
        
        // Phase 1: Get token ranks
        const ranks = await this.getTokenRanksCached(text, ranker);
        
        // Phase 2: Compress ranks with Zipf-based Huffman coding
        const { compressedData, compressionTime } = 
            this.huffmanCompressor.compressRanksWithTiming(ranks, 'zipf');
        
        // Phase 3: Test decompression
        const { decompressedRanks, decompressionTime } = 
            this.huffmanCompressor.decompressRanksWithTiming(compressedData, 'zipf');
        const reconstructedText = await this.getStringFromTokenRanksCached(decompressedRanks, ranker);
        
        // Verify round-trip accuracy
        if (reconstructedText !== text) {
            throw new Error('Round-trip compression failed - reconstructed text doesn\'t match original');
        }
        
        const result = this._createCompressionResult(
            text, compressedData, compressionTime, decompressionTime,
            'llm_ranks_huffman_zipf', modelName, maxContextLength
        );
        
        return { compressedData, result };
    }

    /**
     * Compress text using LLM rank encoding + Huffman coding with Zipf distribution (binary format)
     * @param {string} text - Input text
     * @param {string} modelName - Model name
     * @param {number} maxContextLength - Maximum context length
     * @returns {Object} {compressedData: Buffer, result: CompressionResult}
     */
    async compressWithLLMRanksHuffmanZipfBytes(text, modelName = 'gpt2', maxContextLength = null) {
        const ranker = this._getRanker(modelName, maxContextLength);
        
        // Phase 1: Get token ranks
        const ranks = await this.getTokenRanksCached(text, ranker);
        
        // Phase 2: Compress ranks with Zipf-based Huffman coding (binary format)
        const { compressedData, compressionTime } = 
            this.huffmanCompressor.compressRanksWithTiming(ranks, 'zipf_bytes');
        
        // Phase 3: Test decompression
        const { decompressedRanks, decompressionTime } = 
            this.huffmanCompressor.decompressRanksWithTiming(compressedData, 'zipf_bytes');
        const reconstructedText = await this.getStringFromTokenRanksCached(decompressedRanks, ranker);
        
        // Verify round-trip accuracy
        if (reconstructedText !== text) {
            throw new Error('Round-trip compression failed - reconstructed text doesn\'t match original');
        }
        
        const result = this._createCompressionResult(
            text, compressedData, compressionTime, decompressionTime,
            'llm_ranks_huffman_zipf_bytes', modelName, maxContextLength
        );
        
        return { compressedData, result };
    }

    /**
     * Baseline: compress raw text with gzip
     * @param {string} text - Input text
     * @returns {Object} {compressedData: Buffer, result: CompressionResult}
     */
    async compressWithRawGzip(text) {
        const textBuffer = Buffer.from(text, 'utf8');
        
        // Compression
        const startTime = Date.now();
        const compressedData = await gzip(textBuffer);
        const compressionTime = (Date.now() - startTime) / 1000;
        
        // Test decompression
        const decompStartTime = Date.now();
        const decompressedData = await gunzip(compressedData);
        const decompressionTime = (Date.now() - decompStartTime) / 1000;
        
        // Verify round-trip
        if (decompressedData.toString('utf8') !== text) {
            throw new Error('Round-trip compression failed');
        }
        
        const result = this._createCompressionResult(
            text, compressedData, compressionTime, decompressionTime,
            'raw_gzip'
        );
        
        return { compressedData, result };
    }

    /**
     * Baseline: tokenize then compress token IDs with gzip
     * @param {string} text - Input text
     * @param {string} modelName - Model name
     * @returns {Object} {compressedData: Buffer, result: CompressionResult}
     */
    async compressWithTokenizerGzip(text, modelName = 'gpt2') {
        const ranker = this._getRanker(modelName);
        
        // Tokenize
        const startTime = Date.now();
        const tokenIds = await ranker.tokenize(text);
        const tokenBuffer = Buffer.from(JSON.stringify(tokenIds));
        const compressedData = await gzip(tokenBuffer);
        const compressionTime = (Date.now() - startTime) / 1000;
        
        // Test decompression
        const decompStartTime = Date.now();
        const decompressedTokens = JSON.parse((await gunzip(compressedData)).toString());
        const reconstructedText = await ranker.decode(decompressedTokens);
        const decompressionTime = (Date.now() - decompStartTime) / 1000;
        
        const result = this._createCompressionResult(
            text, compressedData, compressionTime, decompressionTime,
            'tokenizer_gzip', modelName
        );
        
        return { compressedData, result };
    }

    /**
     * Run all compression methods on text and return results
     * @param {string} text - Input text
     * @param {string} modelName - Model name
     * @param {number} maxContextLength - Maximum context length
     * @param {boolean} saveToDb - Whether to save results to database
     * @param {number} batchSize - Batch size for processing
     * @returns {Object} Dictionary mapping method names to results
     */
    async runCompressionBenchmark(text, modelName = 'gpt2', maxContextLength = null, 
                                  saveToDb = true, batchSize = 4) {
        const results = {};
        
        // Define compression methods to test
        const methods = [
            {
                name: 'llm_ranks_gzip',
                func: () => this.compressWithLLMRanks(text, modelName, maxContextLength)
            },
            {
                name: 'llm_ranks_huffman',
                func: () => this.compressWithLLMRanksHuffman(text, modelName, maxContextLength)
            },
            {
                name: 'llm_ranks_huffman_zipf',
                func: () => this.compressWithLLMRanksHuffmanZipf(text, modelName, maxContextLength)
            },
            {
                name: 'llm_ranks_huffman_zipf_bytes',
                func: () => this.compressWithLLMRanksHuffmanZipfBytes(text, modelName, maxContextLength)
            },
            {
                name: 'raw_gzip',
                func: () => this.compressWithRawGzip(text)
            },
            {
                name: 'tokenizer_gzip',
                func: () => this.compressWithTokenizerGzip(text, modelName)
            }
        ];
        
        // Run each method
        for (const method of methods) {
            try {
                const { result } = await method.func();
                results[method.name] = result;
                
                if (saveToDb) {
                    await this._saveResultToDb(result);
                }
            } catch (error) {
                console.error(`${method.name} compression failed:`, error.message);
                results[method.name] = null;
            }
        }
        
        return results;
    }

    /**
     * Save a compression result to the database
     * @param {CompressionResult} result - Result to save
     * @private
     */
    async _saveResultToDb(result) {
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            const query = `
                INSERT INTO compression_results (
                    timestamp, method, original_size, compressed_size, compression_ratio,
                    compression_time, decompression_time, model_name, context_length, 
                    text_sample, metadata
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            `;
            
            const values = [
                new Date().toISOString(),
                result.method,
                result.originalSize,
                result.compressedSize,
                result.compressionRatio,
                result.compressionTime,
                result.decompressionTime,
                result.modelName,
                result.contextLength,
                result.textSample,
                JSON.stringify(result)
            ];
            
            db.run(query, values, function(err) {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            });
        });
    }

    /**
     * Retrieve compression results from database
     * @param {string} method - Filter by method (optional)
     * @returns {Array} List of results
     */
    async getResultsFromDb(method = null) {
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            let query = 'SELECT * FROM compression_results ORDER BY timestamp DESC';
            let params = [];
            
            if (method) {
                query = 'SELECT * FROM compression_results WHERE method = ? ORDER BY timestamp DESC';
                params = [method];
            }
            
            db.all(query, params, (err, rows) => {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve(rows);
                }
            });
        });
    }

    /**
     * Print a formatted comparison table of compression results
     * @param {Object} results - Results dictionary
     */
    printComparisonTable(results) {
        const successfulResults = Object.entries(results).filter(([, result]) => result !== null);
        
        if (successfulResults.length === 0) {
            console.log('No successful compression results to display.');
            return;
        }
        
        console.log('\n' + '='.repeat(100));
        console.log('COMPRESSION COMPARISON RESULTS');
        console.log('='.repeat(100));
        
        // Table header
        console.log(
            'Method'.padEnd(20) + 
            'Original'.padEnd(10) + 
            'Compressed'.padEnd(10) + 
            'Ratio'.padEnd(8) + 
            'Reduction'.padEnd(10) + 
            'Comp (s)'.padEnd(10) + 
            'Decomp (s)'.padEnd(10)
        );
        console.log('-'.repeat(100));
        
        // Sort by compression ratio (best first)
        const sortedResults = successfulResults.sort((a, b) => a[1].compressionRatio - b[1].compressionRatio);
        
        for (const [method, result] of sortedResults) {
            console.log(
                method.padEnd(20) + 
                result.originalSize.toString().padEnd(10) + 
                result.compressedSize.toString().padEnd(10) + 
                result.compressionRatio.toFixed(3).padEnd(8) + 
                `${result.compressionPercentage.toFixed(1)}%`.padEnd(10) + 
                result.compressionTime.toFixed(3).padEnd(10) + 
                result.decompressionTime.toFixed(3).padEnd(10)
            );
        }
        
        console.log('-'.repeat(100));
        
        // Show best performing method
        if (sortedResults.length > 0) {
            const [bestMethod, bestResult] = sortedResults[0];
            console.log(`Best compression: ${bestMethod} (${bestResult.compressionPercentage.toFixed(1)}% reduction)`);
        }
        
        console.log('='.repeat(100));
    }
}

/**
 * Load a text file from the datasets directory
 * @param {string} filename - Filename to load
 * @returns {string} File contents
 */
export async function loadDatasetFile(filename) {
    const fs = await import('fs');
    const path = await import('path');
    
    const filePath = path.join('datasets', filename);
    
    if (!fs.existsSync(filePath)) {
        throw new Error(`Dataset file not found: ${filePath}`);
    }
    
    return fs.readFileSync(filePath, 'utf8');
}